parsers/saarland.py

```diff
@@ -0,0 +1,199 @@
+from urllib.request import *
+
+import json
+import datetime
+
+from pyopenmensa.feed import LazyBuilder
+
+from utils import Parser, Source
+
+
+API_VERSION = "1"
+# please don't copy this key, contact felix@fefrei.de to get your own instead
+API_KEY = "A1RT5M56" + "ibsGaGNZBJQg"
+APP_VERSION = "1"
+SUPPORTED_LANGUAGES = ["de", "en"]
+
+URL_BASE = (
+    'https://mensaar.de/api/%s/%s/%s/%s/'
+    % (API_VERSION, API_KEY, APP_VERSION, SUPPORTED_LANGUAGES[0])
+)
+URL_BASE_DATA = "getBaseData"
+URL_MENU = "getMenu/"
+
+UTC_DATE_STRING = "%Y-%m-%dT%H:%M:%S.%fZ"
+
+ROLES = {
+    's': 'student',
+    'g': 'other',
+    'm': 'employee'
+}
+
+SPECIAL_NOTICES = [
```

Comment:
I don't really like that you need to hard-code certain notice keys here. Maybe we can extend the `getBaseData` call to provide you with the categorization we need? What would we need to add, exactly?

Reply:
I'm building strings to use as notes on the meal object (see lines [58](https://github.com/mswart/openmensa-parsers/pull/74/files/8a2c2afd2ba9c57f951bc042f568b9529dc39cda#diff-91dab909c61ef0b60c3823e035632147R58) and [68](https://github.com/mswart/openmensa-parsers/pull/74/files/8a2c2afd2ba9c57f951bc042f568b9529dc39cda#diff-91dab909c61ef0b60c3823e035632147R68)), because of a pretty limited feed specification (see [openmensa#34](/openmensa/openmensa/issues/34)) with no differentiation between allergic, nutritional or ingredients' information.

Regarding the mensaar.de API, the only differentiation between `notices` is provided by the `isAllergen` value, but most of the `notices` are nouns (and could be formatted as in the lines linked to above), but these "special" ones are adjectives that does not fit within the string. I therefore needed to handle them separately to avoid a `note` like `mit geschwefelt`, for example.

Reply:
I'm not really sure whether it is necessary to extend the API, but it would be a nice improvement to mark these cases differently.

Reply:
Hm, in that case, I think your list is incomplete: “mit kann Spuren von Nüssen beinhalten” isn't right either, but `nu` is not in your exception list.

I'd be happy to extend the API here, but I currently don't see which exact criterion I'd have to set up to help you and others with this problem. Given that English translations aren't currently available anyways, maybe checking whether the notice text starts with a capital letter might be enough? The only exception to this rule seems to be `Aus zum Teil fein zerkleinertem Fleisch`, and I'm removing the capital letter right now 😉

Reply:
> Hm, in that case, I think your list is incomplete: “mit kann Spuren von Nüssen beinhalten” isn't right either, but nu is not in your exception list.

Uuuuh, you're right, I've missed that one! Seems like there are no meals with nuts within the next four days :laughing: 

> Given that English translations aren't currently available anyways, maybe checking whether the notice text starts with a capital letter might be enough?

That would be an alternative, indeed. I'm not sure which one I prefer, though.

> I'd be happy to extend the API here, but I currently don't see which exact criterion I'd have to set up to help you and others with this problem.

Just thinking, if you would introduce another key called `isAdjective` or `isDescriptive` or something similar... or a key for the nouns called `isComponent` or `isIngredient`...?

Reply:
I'm totally open to that, but I haven't found something that sounds totally consistent yet. For example, *kann Spuren von Nüssen beinhalten* is in some sense an ingredient, but for your purpose, you'd clearly want `isIngredient: false`, right?

There are even more corner-cases: What about *Krebstiere, Krusten- und Schalentiere*? These are clearly ingredients, or components, or however you want to call them, but you cannot say *mit Krebstiere, Krusten- und Schalentiere* – you'd need to say *mit Krebstiere**n**, Krusten- und Schalentiere**n***.

Another workaround would be for the API to always return an alternative “descriptive form”, which would never be a single substantive, but always a describing phrase. But that also doesn't feel like a good solution for me.

Reply:
>  For example, kann Spuren von Nüssen beinhalten is in some sense an ingredient, but for your purpose, you'd clearly want isIngredient: false, right?

Is that a possible future example? I didn't get it by the API up to now :thinking: 
I think these kind of cases shouldn't have a `displayName` as _kann Spuren von Nüssen beinhalten_ but only _Nüsse_ and another key, some sort of `isPossibility` or `couldBe`... It would be up to the client to handle this case accordingly.

> What about Krebstiere, Krusten- und Schalentiere?

In this case, I've been lucky because it's marked as `isAllergen` and _enthält Krebstiere, Krusten- und Schalentiere_ works. But that's the problem of the German language and Its accusative forms :see_no_evil: 

> Another workaround would be for the API to always return an alternative “descriptive form”, which would never be a single substantive, but always a describing phrase. But that also doesn't feel like a good solution for me.

Nor to me, to be honest. The best possible solution IMHO would be to restrict the possible values of `displayName` (e.g. to one word) and to abstract other properties to standard keys. But that implies some changes on client-side (e.g. _ohne Schweinefleisch_ should only have _Schweinefleisch_ as `displayName`, while the client has to recognize the `isNegative` key and add an _ohne_ to the displayed form).

Or to keep backward compatibility, you could introduce a new key `name` with the restriction mentioned above and keep the `displayName` key as a sort of "descriptive form" (as it already is in some cases).

Reply:
> Is that a possible future example? I didn't get it by the API up to now 🤔

The `isIngredient` flag was just an example as I'm not sure what the best option is (yet).

> I think these kind of cases shouldn't have a displayName as kann Spuren von Nüssen beinhalten but only Nüsse and another key, some sort of isPossibility or couldBe... It would be up to the client to handle this case accordingly.

Hm, there is some truth to that, but that would be a breaking change (at least semantically), and this text was provided like so from Studentenwerk, so I don't really want to change it.

> In this case, I've been lucky because it's marked as isAllergen and enthält Krebstiere, Krusten- und Schalentiere works. But that's the problem of the German language and Its accusative forms 🙈

You're a lucky guy 😉

>  abstract other properties to standard keys

That sounds good, but it's actually a bit tricky. There are at least two cases (*components* and *properties*, although the separation is sometimes a bit tricky), negation, *and* and *or, ...

Reply:
> The isIngredient flag was just an example as I'm not sure what the best option is (yet).

I was referring to the concrete example of _kann Spuren von Nüssen enthalten_ :smile: 

> Hm, there is some truth to that, but that would be a breaking change (at least semantically), and this text was provided like so from Studentenwerk, so I don't really want to change it.

That's why I thought of leaving this text provided from Studentenwerk as it is for `displayName` and introducing the new (restricted) `name` key.

> There are at least two cases (components and properties, although the separation is sometimes a bit tricky)

In which cases would it be tricky?

> and and *or, ...

I'm not sure what do you mean by _and_ and _or_?

Reply:
> I was referring to the concrete example of kann Spuren von Nüssen enthalten 😄

Oops. Turns out the configuration file provided by Studentenwerk contains both

    "nu": {
        "displayName": {"de": "Schalenfrüchte (Nüsse)", "en": "Nuts"},
        "isAllergen": true
    },

and

    "nu": {
        "displayName": {"de": "kann Spuren von Nüssen beinhalten", "en": "may contain traces of nuts"},
        "isAllergen": false
    },

so we actually only see the former.

> In which cases would it be tricky?

How would you classify *aus zum Teil fein zerkleinertem Fleisch*? In some sense, it's both component and property.

> I'm not sure what do you mean by and and or?

I'm talking about weird notices like *Macadamia-oder-Queenslandnüsse* or *Krebstiere, Krusten- und Schalentiere*. None of these are made up, unfortunately 😟

Reply:
> How would you classify aus zum Teil fein zerkleinertem Fleisch? In some sense, it's both component and property.

Okay, I see. So what do you think about introducing the three different notices groups _components_, _properties_ and _other_, while _other_ would cover these (very few) special cases. So, every _notice_ object would have a key `group` or (`type`) with the only accepted values `component` (or `ingredient`), `property` and `other` (or simply `note`).

> I'm talking about weird notices like Macadamia-oder-Queenslandnüsse or Krebstiere, Krusten- und Schalentiere.

Hmmmm, then why not better leaving these notices' names together? If they are only composed of nouns and _und_/_oder_ :thinking:

Reply:
Hm, that could work, although `other` still feels like cheating. I've asked the IT department of Studentenwerk for their opinion, maybe they have some comments.